File: src/tokenStore.js
'use strict';

const fs = require('fs');
const path = require('path');

// Reads the stored credentials, or null if none / unreadable.
function readToken(tokenPath) {
  try {
    if (!fs.existsSync(tokenPath)) return null;
    const data = JSON.parse(fs.readFileSync(tokenPath, 'utf8'));
    if (data && typeof data.token === 'string' && data.token.length > 0) {
      return { agentId: data.agentId ?? null, token: data.token };
    }
    return null;
  } catch {
    return null;
  }
}

// Persists the token with restrictive (owner-only) permissions. chmod is applied
// explicitly because writeFile's mode only takes effect when creating the file.
function saveToken(tokenPath, { agentId, token }) {
  fs.mkdirSync(path.dirname(tokenPath), { recursive: true });
  const body = `${JSON.stringify({ agentId, token }, null, 2)}\n`;
  fs.writeFileSync(tokenPath, body, { mode: 0o600 });
  fs.chmodSync(tokenPath, 0o600);
}

module.exports = { readToken, saveToken };

File: blueeye-agent-go/internal/tokenstore/tokenstore.go
// Package tokenstore reads and writes the agent credential exactly as the Node
// agent does (GO-REWRITE-AUDIT.md §5): a JSON document {agentId, token} — NOT a
// bare token string — persisted at mode 0600 with an explicit chmod after write.
package tokenstore

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
)

// Credentials is the on-disk credential shape. agentId may be null in the file;
// it is represented as a pointer so a missing/null value round-trips.
type Credentials struct {
	AgentID *int64 `json:"agentId"`
	Token   string `json:"token"`
}

// Read returns the stored credentials, or (nil, nil) when there is no usable
// token (missing file, unreadable, bad JSON, or empty token) — matching Node's
// readToken which returns null in all those cases.
func Read(path string) (*Credentials, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, nil // unreadable is treated as "not enrolled", like Node
	}
	var c Credentials
	if err := json.Unmarshal(data, &c); err != nil {
		return nil, nil
	}
	if c.Token == "" {
		return nil, nil
	}
	return &c, nil
}

// Save writes the credentials with owner-only permissions. It creates the parent
// directory and applies chmod 0600 explicitly (the file mode on create is not
// enough if the file already existed), matching Node's saveToken.
func Save(path string, c Credentials) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	body, err := json.MarshalIndent(c, "", "  ")
	if err != nil {
		return err
	}
	body = append(body, '\n')
	if err := os.WriteFile(path, body, 0o600); err != nil {
		return err
	}
	return os.Chmod(path, 0o600)
}
